gtk/gtkprintoperation-portal.c

```diff
@@ -0,0 +1,597 @@
+    }
+
+  g_variant_get (parameters, "(u@a{sv})", &response, &options);
+
+  portal->do_print = (response == 0);
+
+  if (portal->do_print)
+    {
+      portal->result = GTK_PRINT_OPERATION_RESULT_APPLY;
+      GVariant *v;
+      GtkPrintSettings *settings;
+      GtkPageSetup *page_setup;
+      GtkPrinter *printer;
+      char *filename;
+      char *uri;
+      int fd;
+
+      v = g_variant_lookup_value (options, "settings", G_VARIANT_TYPE_VARDICT);
+      settings = gtk_print_settings_new_from_gvariant (v);
+      g_variant_unref (v);
+
+      v = g_variant_lookup_value (options, "page-setup", G_VARIANT_TYPE_VARDICT);
+      page_setup = gtk_page_setup_new_from_gvariant (v);
+      g_variant_unref (v);
+
+      g_variant_lookup (options, "token", "u", &portal->token);
+
+      printer = find_file_printer ();
+
+      fd = g_file_open_tmp ("gtkprintXXXXXX", &filename, NULL);
+      uri = g_strconcat ("file://", filename, NULL);
+      gtk_print_settings_set (settings, GTK_PRINT_SETTINGS_OUTPUT_URI, uri);
+      g_free (uri);
+      close (fd);
+
+      finish_print (portal, printer, page_setup, settings);
+      g_free (filename);
+    }
+  else
+    portal->result = GTK_PRINT_OPERATION_RESULT_CANCEL;
+
+  if (portal->loop)
+    g_main_loop_quit (portal->loop);
+}
+
+static void
+print_called (GObject *source,
+              GAsyncResult *result,
+              gpointer data)
+{
+  PortalData *portal = data;
+  GError *error = NULL;
+  char *handle = NULL;
+
+  if (!gxdp_print_call_print_finish (portal->proxy, &handle, result, &error))
+    {
+      g_warning ("Error: %s", error->message);
+      g_error_free (error);
+      g_main_loop_quit (portal->loop);
+      return;
+    }
+
+  portal->print_response_signal_id =
+    g_dbus_connection_signal_subscribe (g_dbus_proxy_get_connection (G_DBUS_PROXY (portal->proxy)),
+                                        "org.freedesktop.portal.Desktop",
+                                        "org.freedesktop.portal.Request",
+                                        "Response",
+                                        handle,
+                                        NULL,
+                                        G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
+                                        print_response,
+                                        portal, NULL);
+
+  g_free (handle);
+}
+
+GtkPrintOperationResult
+gtk_print_operation_portal_run_dialog (GtkPrintOperation *op,
+                                       gboolean           show_dialog,
+                                       GtkWindow         *parent,
+                                       gboolean          *do_print)
+{
+  GtkPrintOperationPrivate *priv = op->priv;
+  GVariant *settings;
+  GVariant *setup;
+  GVariantBuilder opt_builder;
+  GVariant *options;
+  guint signal_id;
+  PortalData *portal;
+  GtkPrintOperationResult result;
+
+  portal = g_new0 (PortalData, 1);
+  portal->proxy = get_portal_proxy ();
```

Comment:
I think we should check for `proxy` being `NULL` and return an error (also below)